slowapi/middleware.py

```diff
@@ -54,3 +58,89 @@ async def dispatch(
             response = limiter._inject_headers(response, request.state.view_rate_limit)
             return response
         return await call_next(request)
+
+
+class SlowAPIASGIMiddleware:
+    def __init__(self, app: ASGIApp) -> None:
+        self.app = app
+        self.error_response: Optional[Response] = None
+        self.initial_message: Message = {}
+        self.inject_headers = False
+
+    async def send_wrapper(self, message: Message):
+        if message["type"] == "http.response.start":
+            # do not send the http.response.start message now, so that we can edit the headers
+            # before sending it, based on what happens in the http.response.body message.
+            self.initial_message = message
+
+        elif message["type"] == "http.response.body":
+            if self.error_response:
+                self.initial_message["status"] = self.error_response.status_code
+
+            if self.inject_headers:
+                headers = MutableHeaders(raw=self.initial_message["headers"])
+                headers = self.limiter._inject_asgi_headers(
+                    headers, self.request.state.view_rate_limit
+                )
+
+            # send the http.response.start message just before the http.response.body one,
+            # now that the headers are updated
+            await self.send(self.initial_message)
+            await self.send(message)
+
+    async def __call__(self, scope: Scope, receive: Receive, send: Send) -> None:
+        if scope["type"] != "http":
+            await self.app(scope, receive, send)
+            return
+
+        self.send = send
+
+        _app: Starlette = scope["app"]
+        limiter: Limiter = _app.state.limiter
+        handler = None
+
+        if not limiter.enabled:
+            return await self.app(scope, receive, self.send)
+
+        for route in _app.routes:
+            match, _ = route.matches(scope)
+            if match == Match.FULL and hasattr(route, "endpoint"):
+                handler = route.endpoint  # type: ignore
+
+        # if we can't find the route handler
+        if handler is None:
+            return await self.app(scope, receive, self.send)
+
+        name = "%s.%s" % (handler.__module__, handler.__name__)
```

Comment:
Any particular reason not to use the f-string notation here? I thought it was faster pretty much all the time. Personally, I also find it much more readable.

Reply:
not really, just a refactor of the existing code, didn't touch this line